### PR TITLE
Run send mission in separate thread

### PR DESCRIPTION
**Pull request by vetlek**

### CODE REVIEW

**Review by Eddasol — APPROVED** [review]
LGTM

File: src/isar/state_machine/states/send.py [hunk 1]
```diff
@@ -26,70 +40,84 @@ def start(self):
 
     def stop(self):
         self.send_failure_counter = 0
+        if self.send_thread:
+            self.send_thread.join()
```
Comment (oysand) [inline src/isar/state_machine/states/send.py:44]: Could we risk the code hanging on this step if something is holding up the thread?

Comment (oysand) [reply src/isar/state_machine/states/send.py:44]: Well, that is the purpose of join come to think on it... What I mean is that is there a case where the code could hang indefinitely, if so, should we implement some sort of timeout?

File: src/isar/state_machine/states/send.py [hunk 2]
```diff
@@ -26,70 +40,84 @@ def start(self):
...
-                    f"Failed to send mission after {self.send_failure_counter_limit} attempts. Cancelling mission."
+            if not self.send_thread:
+                self.state_machine.status.current_mission_step = (
+                    self._get_current_mission()
                 )
-                next_state: States = States.Cancel
-                break
+                self.send_thread = threading.Thread(
+                    target=send_mission_thread,
+                    args=(
```
Comment (oysand) [inline src/isar/state_machine/states/send.py:68]: Could we list which input each of these would be, for instance: schedule_step_func=self.state_machine.robot.schedule_step?

File: src/isar/state_machine/states/send.py [hunk 3]
```diff
@@ -26,70 +40,84 @@ def start(self):
...
-                    computed_joints
-                )
-        else:
-            send_success = False
+        return self.state_machine.status.mission_schedule.mission_steps[0]
+
 
-        if self.state_machine.should_send_status():
-            self.state_machine.send_status()
+def send_mission_thread(
```
Comment (oysand) [inline src/isar/state_machine/states/send.py:112]: Would it be better to have this function as part of the class?

**Review by oysand — CHANGES_REQUESTED** [review]
See comments/questions

File: src/isar/services/utilities/threaded_request.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class ThreadedRequest:
+    def __init__(self, request_func: Any) -> None:
+        self._thread: Optional[Thread] = None
+        self._request_func: Any = request_func
+        self._output: Any = None
+        self._output_lock: Lock = Lock()
+
+    def start_thread(self, *args) -> bool:
```
Comment (oysand) [inline src/isar/services/utilities/threaded_request.py:13]: Should we specify which args one can input?

File: src/isar/state_machine/states/send.py [hunk 5]
```diff
@@ -26,70 +29,71 @@ def start(self):
...
-                next_state: States = States.Cancel
-                break
+                self.send_thread = ThreadedRequest(
+                    self.state_machine.robot.schedule_step
+                )
+                self.send_thread.start_thread(
+                    self.state_machine.status.current_mission_step
+                )
+
+            if not self.send_thread.is_thread_alive():
```
Comment (oysand) [inline src/isar/state_machine/states/send.py:61]: Handle thread specific requirements inside ThreadRequest. Could try try except?

File: src/isar/state_machine/states/send.py [hunk 6]
```diff
@@ -26,70 +33,75 @@ def start(self):
...
+                    self.state_machine.status.current_mission_step
+                )
+            try:
+                (
+                    send_success,
+                    mission_instance_id,
+                    computed_joints,
+                ) = self.send_thread.get_output()
+            except ThreadedRequestNotFinishedError:
+                continue
```
Comment (oysand) [inline src/isar/state_machine/states/send.py:71]: Consider a short sleep?

**Review by Christdej — APPROVED** [review]

**Pull request merged**